packages/xsnap/src/xsnap.c

```diff
@@ -1430,6 +1456,46 @@ static void fx_issueCommand(xsMachine *the)
 	xsResult = xsArrayBuffer(buf, len);
 }
 
+
+void adjustSpaceMeter(txMachine* the, txSize theSize)
+{
+	txSize previous = the->allocatedSpace;
+	the->allocatedSpace += theSize;
+	if (the->allocatedSpace > the->allocationLimit ||
+		// overflow?
+		the->allocatedSpace < previous) {
+		fxAbort(the, XS_NOT_ENOUGH_MEMORY_EXIT);
```

Comment:
let's add a comment here to remind ourselves that our `fxAbort` must absolutely `c_exit` in response to `XS_NOT_ENOUGH_MEMORY_EXIT`.. unlike the other exit codes, this one happens while the engine is in a delicate state and something like `fxExitToHost` would leave the engine corrupted

Reply:
As per Peter, we should fxabort with a different error code if allocation meter exhausted, so that we can exit to host for that, and still abort the process for the other memory error.

Reply:
> let's add a comment here to remind ourselves that our `fxAbort` must absolutely `c_exit`

I added the comment, but this seems like an odd place for it; I put it on fxAbort.

Reply:
> As per Peter, we should fxabort with a different error code if allocation meter exhausted, so that we can exit to host for that, and still abort the process for the other memory error.

I think that's what we're doing. I wonder if I'm missing your point, @michaelfig .

Reply:
> I think that's what we're doing. I wonder if I'm missing your point, @michaelfig .

We mustn't use `XS_NOT_ENOUGH_MEMORY_EXIT` to indicate an allocation meter overflow, since that is used for engine-invariant-failure and is mandatorily engine-terminating.

Instead, use something to disambiguate an engine-level failure (`XS_NOT_ENOUGH_MEMORY_EXIT`) and an allocation metering exhaustion like (a new) `XS_TOO_MUCH_ALLOCATION`.  `XS_TOO_MUCH_ALLOCATION` would just be advisory, and `fxAbort` can legitimately choose to do an exit-to-host without terminating the engine, just like `XS_TOO_MUCH_COMPUTATION` currently does.

Reply:
OK, I see your point, @michaelfig ; it looks like part of the non-fatal metering faults that we're postponing in #2979 , ok?

Reply:
> OK, I see your point, @michaelfig ; it looks like part of the non-fatal metering faults that we're postponing in #2979 , ok?

Okay, fine to postpone!  I should have known that @warner had written this up somewhere.